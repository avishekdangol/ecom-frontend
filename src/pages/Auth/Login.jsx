import {
  Input, Checkbox, Button, Spin,
} from 'antd';
import { LoadingOutlined } from '@ant-design/icons';
import { BiUser, BiLock } from 'react-icons/bi';
import { FaEye, FaEyeSlash } from 'react-icons/fa';
import { Link, useNavigate } from 'react-router-dom';
import {
  Formik, Form,
} from 'formik';
import { useState } from 'react';
import LoginSchema from './validations/LoginSchema';
import { showErrorNotification } from '@/utils/Toasts';
import { useAuth } from '@/utils/AuthContext';
import jwt from '@/auth/useJwt';

function Login() {
  const { login } = useAuth();
  const navigate = useNavigate();
  const navigateHome = () => {
    navigate('/');
  };

  const [processing, setProcessing] = useState(false);
  const [passwordType, setPasswordType] = useState('password');

  const isProcessing = <LoadingOutlined style={{ fontSize: 24 }} spin />;
  const togglePasswordType = () => {
    setPasswordType(passwordType === 'password' ? 'text' : 'password');
  };

  return (
    <Formik
      initialValues={{
        email: '',
        password: '',
        remember: false,
      }}
      validationSchema={LoginSchema}
      onSubmit={(values) => {
        setProcessing(true);
        jwt.login(values).then((response) => {
          // eslint-disable-next-line camelcase
          const { user, access_token } = response.data;
          localStorage.setItem('userData', JSON.stringify(user));
          localStorage.setItem('accessToken', access_token);
          login();

          navigateHome();
        }).catch(({ response }) => {
          showErrorNotification('Login Failed!', response.data.message);
        }).finally(() => {
          setProcessing(false);
        });
      }}
    >
      {({
        values, touched, errors, handleChange, handleSubmit,
      }) => (
        <Form>
          {processing}
          <div className="mb-4">
            <Input
              id="email"
              name="email"
              type="email"
              value={values.email}
              placeholder="Email"
              prefix={<BiUser />}
              onPressEnter={handleSubmit}
              onChange={handleChange}
            />
            <p className="mb-0">
              {touched.email && errors.email && <small className="text-red-500">{errors.email}</small>}
            </p>
          </div>

          <div className="mb-4">
            <Input
              id="password"
              name="password"
              type={passwordType}
              value={values.password}
              placeholder="Password"
              prefix={<BiLock />}
              suffix={passwordType === 'password'
                ? (
                  <FaEye
                    className="cursor-pointer"
                    onClick={togglePasswordType}
                  />
                ) : (
                  <FaEyeSlash
                    className="cursor-pointer"
                    onClick={togglePasswordType}
                  />
                )}
              onPressEnter={handleSubmit}
              onChange={handleChange}
            />
            <p className="mb-0">
              {touched.password && errors.password && <small className="text-red-500">{errors.password}</small>}
            </p>
          </div>

          <Checkbox
            id="remember"
            name="remember"
            checked={values.remember}
            onChange={handleChange}
          >
            Remember Me
          </Checkbox>

          <Button
            type="primary"
            className="primary-btn block my-6 w-full"
            disabled={processing}
            onClick={handleSubmit}
          >
            {
              processing
                ? <Spin indicator={isProcessing} />
                : 'Login'
            }
          </Button>

          <p className="text-xs text-center">
            Don&apos;t have an account?
            {' '}
            <Link
              to="/register"
              className="inline text-blue-700 underline hover:text-blue-500 hover:no-underline"
            >
              Create one.
            </Link>
          </p>
        </Form>
      )}
    </Formik>
  );
}

export default Login;
