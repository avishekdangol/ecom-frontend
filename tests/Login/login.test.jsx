import { render } from '@root/tests/testUtils';
import Login from '@/pages/Auth/Login';

describe('Login', () => {
  test('snapshot', () => {
    const { asFragment } = render(<Login />);
    expect(asFragment()).toMatchSnapshot();
  });
});
