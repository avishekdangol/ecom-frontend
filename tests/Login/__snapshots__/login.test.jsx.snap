// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login snapshot 1`] = `
<DocumentFragment>
  <form
    action="#"
  >
    <div
      class="mb-4"
    >
      <span
        class="ant-input-affix-wrapper css-dev-only-do-not-override-1g853jt"
      >
        <span
          class="ant-input-prefix"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2a5 5 0 1 0 5 5 5 5 0 0 0-5-5zm0 8a3 3 0 1 1 3-3 3 3 0 0 1-3 3zm9 11v-1a7 7 0 0 0-7-7h-4a7 7 0 0 0-7 7v1h2v-1a5 5 0 0 1 5-5h4a5 5 0 0 1 5 5v1z"
            />
          </svg>
        </span>
        <input
          class="ant-input css-dev-only-do-not-override-1g853jt"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </span>
      <p
        class="mb-0"
      />
    </div>
    <div
      class="mb-4"
    >
      <span
        class="ant-input-affix-wrapper css-dev-only-do-not-override-1g853jt"
      >
        <span
          class="ant-input-prefix"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 2C9.243 2 7 4.243 7 7v2H6c-1.103 0-2 .897-2 2v9c0 1.103.897 2 2 2h12c1.103 0 2-.897 2-2v-9c0-1.103-.897-2-2-2h-1V7c0-2.757-2.243-5-5-5zM9 7c0-1.654 1.346-3 3-3s3 1.346 3 3v2H9V7zm9.002 13H13v-2.278c.595-.347 1-.985 1-1.722 0-1.103-.897-2-2-2s-2 .897-2 2c0 .736.405 1.375 1 1.722V20H6v-9h12l.002 9z"
            />
          </svg>
        </span>
        <input
          class="ant-input css-dev-only-do-not-override-1g853jt"
          name="password"
          placeholder="Password"
          type="password"
          value=""
        />
        <span
          class="ant-input-suffix"
        >
          <svg
            class="cursor-pointer"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </span>
      </span>
      <p
        class="mb-0"
      />
    </div>
    <div
      class="flex justify-between items-center"
    >
      <label
        class="ant-checkbox-wrapper text-xs css-dev-only-do-not-override-1g853jt"
      >
        <span
          class="ant-checkbox ant-wave-target css-dev-only-do-not-override-1g853jt"
        >
          <input
            class="ant-checkbox-input"
            name="remember"
            type="checkbox"
          />
          <span
            class="ant-checkbox-inner"
          />
        </span>
        <span>
          Remember Me
        </span>
      </label>
      <a
        class="text-xs text-blue-700 underline hover:text-blue-500 hover:no-underline"
        href="/forgot-password"
      >
        Forgot Password?
      </a>
    </div>
    <button
      class="ant-btn css-dev-only-do-not-override-1g853jt ant-btn-primary primary-btn block my-6 w-full"
      type="button"
    >
      <span>
        Login
      </span>
    </button>
    <p
      class="text-xs text-center"
    >
      Don't have an account? 
      <a
        class="inline text-blue-700 underline hover:text-blue-500 hover:no-underline"
        href="/register"
      >
        Create one.
      </a>
    </p>
  </form>
</DocumentFragment>
`;
